### PR TITLE
Update SmallDirS to include nucleation; Calculate nucleation fraction and print to log

### PR DESCRIPTION
**Pull request by MattRolchigo**
SmallDirS now includes nucleation - the problem is tested via a new unit test that runs the problem and checks that the volume fraction of nucleated grains is within some tolerance. To do this, the volume fraction of nucleated grains is created as a return value of type float for the simulations, and the calculation of the volume fraction of nucleated grains was moved out of `PrintGrainMisorientations` (it is now calculated in a separate subroutine and called during the final collection of grain ID values)

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -37,7 +37,11 @@ int main(int argc, char *argv[]) {
         else {
             // Run CA code using reduced temperature data format
             std::string InputFile = argv[1];
-            RunProgram_Reduced(id, np, InputFile);
+            float VolFractionNucleated = RunProgram_Reduced(id, np, InputFile);
```
[the remark this answers is not in the document]
Comment (MattRolchigo) [reply src/main.cpp:40]: Yeah that might be less disruptive to the code structure to just have this info printed to the log

File: examples/Inp_SmallDirSolidification.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
        "Nz": 50
```
[the remark this answers is not in the document]
Comment (MattRolchigo) [reply examples/Inp_SmallDirSolidification.json:11]: Yes but we'd have to increase the nucleation/surface grain density, and I didn't want to change too much about this test problem all at once

File: src/CAfunctions.cpp [hunk 3]
```diff
@@ -164,3 +164,54 @@ ViewF_H MisorientationCalc(int NumberOfOrientations, ViewF_H GrainUnitVector, in
...
+            // Is this Y coordinate in the halo region? If so, do not increment counter
+            bool InHaloRegion = false;
+            if (((RankY == 0) && (!AtSouthBoundary)) || ((RankY == MyYSlices - 1) && (!AtNorthBoundary)))
+                InHaloRegion = true;
+            if ((LayerID(D3D1ConvPosition) != -1) && (!InHaloRegion))
+                update++;
+        },
+        MeltedCells_Local);
+
+    // For interior cells, add the number of cells with sub-zero grain IDs (TODO: could be combined with the above loop
```
Comment (streeve) [inline src/CAfunctions.cpp:188]: The reduction can be combined really easily with variadic inputs/functor signature - we do this in the `IntermediateOutputAndCheck` reduction

File: src/CAfunctions.cpp [hunk 4]
```diff
@@ -164,3 +164,41 @@ ViewF_H MisorientationCalc(int NumberOfOrientations, ViewF_H GrainUnitVector, in
...
+float calcVolFractionNucleated(int id, int nx, int MyYSlices, int LocalDomainSize, ViewI LayerID, ViewI GrainID,
+                               bool AtNorthBoundary, bool AtSouthBoundary) {
+
+    // For interior cells, add the number of cells that underwent melting/solidification and the number of cells with
+    // sub-zero grain IDs
+    int MeltedCells_Local = 0;
+    int NucleatedGrainCells_Local = 0;
+    Kokkos::parallel_reduce(
+        "NumSolidifiedCells", LocalDomainSize,
+        KOKKOS_LAMBDA(const int D3D1ConvPosition, int &update_meltcount, int &update_nucleatecount) {
```
Comment (streeve) [inline src/CAfunctions.cpp:178]: Would prefer `melt_count`, etc.

**Review by streeve — APPROVED** [review]

**Pull request merged**